### PR TITLE
Idl + misc 

### PR DESCRIPTION
**Pull request by Supercip971**
added: 
- a basic idl 
- libc: ctypes
- fmt: char support 
- scan: scan_skip_until
- fix
- test: assert for string
- io write/read for buffer

### CODE REVIEW

File: sources/kernel/.build.mk [hunk 1]
```diff
@@ -5,11 +5,13 @@ KERNEL_C_SRC+=$(wildcard sources/kernel/*.c) \
...
+			  $(wildcard sources/libraries/ansi/*.c) \
 			  $(wildcard sources/libraries/ubsan/*.c) \
 			  $(wildcard sources/libraries/brutal/io/fmt.c) \
 			  $(wildcard sources/libraries/brutal/io/print.c) \
 			  $(wildcard sources/libraries/brutal/io/scan.c) \
 			  $(wildcard sources/libraries/brutal/io/write.c) \
+			  $(wildcard sources/libraries/brutal/io/buffer.c) \
 			  $(wildcard sources/libraries/brutal/mem/*.c) \
 			  $(wildcard sources/libraries/brutal/text/*.c) \
 			  $(wildcard sources/libraries/brutal/sync/*.c) \
```
Comment (sleepy-monax) [inline sources/kernel/.build.mk:17]: Why are we including that into the kernel? Why the whole ANSI C library?

File: sources/bid/bid.c [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        }
+    }
+
+    if (end == 0 && start != 0)
+    {
+        end = input.len;
+    }
+
+    return str_cast_n(end - start, input.buffer + start);
+}
```
Comment (sleepy-monax) [inline sources/bid/bid.c:59]: We should move that into the scan object, like scan_current_location(struct scan* self);

File: sources/libraries/bid/bid.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
+#pragma once
+#include <brutal/base.h>
+#include <brutal/ds.h>
+#include <brutal/io/scan.h>
+
+enum bid_error_type
+{
+    BID_ERROR_TYPE_INVALID,
+    BID_ERROR_TYPE_UNEXPECTED_TOKEN,
+};
```
Comment (sleepy-monax) [inline sources/libraries/bid/bid.h:10]: This type should not exist, we should use Error instead

File: sources/libraries/bid/bid.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+struct bid_buffer_position
+{
+    int offset;
+};
+
+struct bid_error
+{
+    enum bid_error_type type;
+    struct bid_buffer_position pos;
+    Str message;
```
Comment (sleepy-monax) [inline sources/libraries/bid/bid.h:21]: No need for message here if we use Error

File: sources/libraries/bid/bid.h [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+struct bid
+{
+    Str in_data;
+
+    struct bid_ast_node *root_ast;
+    struct scan scanner;
+    int _current_scanned_token_cursor;
+};
+
+typedef Result(struct bid_error, MonoState) BidParseResult;
```
Comment (sleepy-monax) [inline sources/libraries/bid/bid.h:43]: We should use MaybeError instead

File: sources/libraries/bid/bid.h [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+typedef Result(struct bid_error, struct bid) BidResult;
```
Comment (sleepy-monax) [inline sources/libraries/bid/bid.h:44]: We should use Error

File: sources/libraries/bid/bid.h [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+BidResult init_bid(Str idl_in);
+
+void destroy_bid(struct bid *in);
```
Comment (sleepy-monax) [inline sources/libraries/bid/bid.h:48]: This should be bid_init() and bid_deinit()

File: sources/libraries/bid/c_convert/bid_convert_c.c [hunk 8]
```diff
@@ -0,0 +1,50 @@
+#include <bid/bid.h>
+#include <bid/c_convert/bid_convert_c.h>
+#include <bid/c_convert/bid_convert_errors.h>
+#include <bid/c_convert/bid_convert_header.h>
+#include <bid/c_convert/bid_convert_message_type.h>
+#include <bid/c_convert/bid_convert_method.h>
+#include <bid/c_convert/bid_convert_type.h>
+#include <bid/c_convert/bid_convert_typedef.h>
```
Comment (sleepy-monax) [inline sources/libraries/bid/c_convert/bid_convert_c.c:8]: All these should be a single header file like <bid/c_convert.h> most of them contains a single line a declaration anyway

File: sources/libraries/bid/c_convert/bid_convert_c.c [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+#include <bid/parsing/bid_ast.h>
+#include <brutal/alloc.h>
+#include <brutal/base.h>
+#include <brutal/log.h>
+#include "brutal/text/str.h"
+
```
Comment (sleepy-monax) [inline sources/libraries/bid/c_convert/bid_convert_c.c:14]: You should include <brutal/text.h> and fix you IDE please, we already talk about that in previous PR

**Review by sleepy-monax — COMMENTED** [review]
Good work :+1: 

The structure of this subproject is way more complicated than it needs to be. This makes it hard to navigate. IMO this project should follow the following structure.

```
bid/
bid/ast.c
bid/ast.h
bid/parser.c
bid/parser.h
bid/generator-c.c
bid/generator-c.h
```

File: sources/kernel/.build.mk (hunk 1 above)
Comment (Supercip971) [reply sources/kernel/.build.mk:17]: for ctypes but I should not use the whole ANSI C lib :x

**Review by sleepy-monax — APPROVED** [review]

**Pull request merged**